### ISSUE TITLE
Group and Ungroup table rows based on column name

### ISSUE THREAD

**Issue by rajeshbtlit** [wenzhixin/bootstrap-table#694]
In the boot strap table, group and un-group table rows based on column name. For example in a given table below grouping is done using column  workflow name

![table](https://cloud.githubusercontent.com/assets/11829880/7018059/82860232-dd1d-11e4-9eb3-ea64f15b0000.png)
![groupby](https://cloud.githubusercontent.com/assets/11829880/7018062/8cff1604-dd1d-11e4-9830-39d1a2f3ffef.png)

**Comment by djhvscf** [comment]
Ref to: https://github.com/wenzhixin/bootstrap-table/issues/73

**Comment by rajeshbtlit** [comment]
Hi Thanks, I will try it out, can we collapse and expand the rows after grouping.

**Comment by djhvscf** [comment]
Right now it is just an idea.. I'm working on it

**Comment by rajeshbtlit** [comment]
I really appreciate your response, is this already in the repo , if so can we just use, like for example,
We have a button group, and on click of button can we just call
$('#groupWorkflow').on('click',function() {
		                   		$('#tableWorkflow').bootstrapTable({
		                	        groupBy: true,
		                	        groupFields: ['wfdName']
		                	         });
		                	});

**Comment by kryptonianson** [comment]
This is the exact implementation I also am looking for. We have something similar in another web technology component we use and we really need to dump it quick. Please let me know if there is anything I can do to help move this along. Thank you for your hard work and efforts.

**Comment by djhvscf** [comment]
See this: https://github.com/djhvscf/bootstrap-table/blob/master/src/extensions/group-by/bootstrap-table-group-by.js 
we are working on it

**Comment by kryptonianson** [comment]
looks like you are still active on it. great news for us! thank you so much for all your hard work.

**Comment by djhvscf** [comment]
Yes, it is working well but we have some issues related to sort events and css style

**Comment by djhvscf** [comment]
See this http://jsfiddle.net/djhvscf/e3nk137y/2258/ and this https://github.com/wenzhixin/bootstrap-table/pull/1073 @kryptonianson @rajeshbtlit

**Comment by Knoxvillekm** [comment]
With group-by extension table lost functionality (sort, pagination, footter, etc).
How about this - http://jsfiddle.net/Knoxvillekm/qbyxfamr/

This decision needs to be expanded, but it is functional.
It does not require jquery.treetable.

**Comment by Knoxvillekm** [comment]
With group-by extension table lost functionality (sort, pagination, footer, etc).

some fix - http://jsfiddle.net/Knoxvillekm/n9fa0t75/
This decision needs to be expanded, but it is more functional.
It does not require jquery.treetable.

![group](https://cloud.githubusercontent.com/assets/11072464/9330236/a9d4da5a-45c0-11e5-9fcf-4240e42a1169.png)

**Comment by gowthamsathiya** [comment]
@Knoxvillekm This works great.

**Comment by djhvscf** [comment]
Can you open a new PR with these changes? @Knoxvillekm or @gowthamsathiya

**Comment by rajeshbtlit** [comment]
Thanks @djhvscf  and @gowthamsathiya .
Your work is highly appreciated.

**Comment by Knoxvillekm** [comment]
@djhvscf, this is alpha version.

**Comment by deepfriedbrain** [comment]
@Knoxvillekm The table works great. Thank you!

Coming from Java programming background, I'm relatively new to jQuery. Is there a way I can make the group-by-field selection dynamic? Like a small drop down menu that allows the user to select/change the group-by-field dynamically?

Any help or pointers would be appreciated.

Thanks again.

**Comment by deepfriedbrain** [comment]
A few suggestions:

1. An up/down arrow representing the collapsed/expanded state in the group headers.
2. A default value to set the initial state of the groups to collapsed/expanded.
3. An option to select the group-by-field dynamically (as stated in my previous comment).

Thanks.

**Comment by gowthamsathiya** [comment]
@deepfriedbrain  you could do group-by-field dynamic selection as follows if you have list of items to group by in a dropdown defined by yourself using bootstrap table refresh option.

```
<select name="dropdown" id="groupby">
      <option value="column1" >column1</option>
      <option value="column2">column2</option>
      <option value="column3">column3</option>
</select>


<script>
     var $table = $('#table');
     $(function () {
          $('select[name="dropdown"]').change(function () {
          $table.bootstrapTable('refreshOptions', {
          groupBy: true,
          groupByField: $(this).val()
          });
          });
     });
</script>
```

**Comment by deepfriedbrain** [comment]
@gowthamsathiya This is wonderful. This solution works perfectly for me! I have been learning to code with jQuery and also going through the Bootstrap table documentation, and beginning to get a hang of it. 

I really appreciate your help and all the good work that you guys have been doing. Thanks so much.

**Comment by deepfriedbrain** [comment]
@Knoxvillekm and @gowthamsathiya I did notice one issue with the group by table ( http://jsfiddle.net/Knoxvillekm/n9fa0t75/ ). If I specify a data-height on the table, then all the column headers, but the first, disappear. Please look into when you get a chance.

**Comment by deepfriedbrain** [comment]
I found a workaround to the header issue described in my previous comment by changing the bootstrap-table.js. 

Original:

`
that.$header_.find(sprintf('th[data-field="%s"]', visibleFields[index]))
                .find('.fht-cell').width($this.innerWidth());
`

Modified:

`
that.$header_.find(sprintf('th[data-field="%s"]', visibleFields[index]))
                .find('.fht-cell').width();
`

However, I have no idea about the other problems that it might introduce.

**Comment by deepfriedbrain** [comment]
Guys, sorry for the multiple comments, but I found one more issue with @Knoxvillekm 's table.

If you change data-group-by to false, there's an error in the console:

Uncaught TypeError: Cannot read property 'forEach' of undefined 

It is from this function:

`
    BootstrapTable.prototype.updateSelected = function () {
        if (!initBodyCaller) {
            _updateSelected.apply(this, Array.prototype.slice.apply(arguments));

            this.$selectGroup.forEach(function(item) {
                var checkGroup = item.item.filter(':enabled').length ===
                    item.item.filter(':enabled').filter(':checked').length;

                item.group.prop('checked', checkGroup);
            });
        }
    };
`

**Comment by wenzhixin** [comment]
@Knoxvillekm, I am look forward to your pull request:)

**Comment by endra89** [comment]
bug when multiple table please see  @djhvscf  @Knoxvillekm  @wenzhixin : 
http://jsfiddle.net/endra89/341n34LL/5/
table1 grouing enable
table2 and table 3 no grouping

**Comment by Knoxvillekm** [comment]
@wenzhixin, no idea how to make pull request

**Comment by wenzhixin** [comment]
@Knoxvillekm https://github.com/wenzhixin/bootstrap-table/blob/master/CONTRIBUTING.md#pull-requests

**Comment by evilcve** [comment]
Very nice work here @Knoxvillekm. Got a small addition to your code, when using data-detail-view="true" i added this under the 'this.columns.forEach' function so that the groupby blue color spans across all the columns. Without it the last column remains white, small thing i know, but it may help someone. Again nice work on this.

if (this.options.detailView == true) {
      visibleColumns++;
};

**Comment by KulanthaivelChinnadurai** [comment]
This works great. Thanks for your support.
For my requirement I need to Append/Remove groups dynamically. 
Need your support guys.

**Comment by dctdct** [comment]
Hey, i added the attributes to my table but i did not work:    data-single-select="true"
                           data-groupBy="true"
                           data-groupByField="canEdit"

**Comment by endra89** [comment]
hello, need your help @Knoxvillekm, @wenzhixin 
here my problem : https://jsfiddle.net/endra89/swxpkfc2/1/
need dynamic check box, but still grouped like example, 
please help, thank you.

**Comment by nando-San** [comment]
Hi guys, i found error in group by v2, only works in sigle table,  multiple tables cause error when  click in pagination.

[(http://jsfiddle.net/rma15xzm/2/)]

i hope help me.

**Comment by Lutex** [comment]
Hello guys, @Knoxvillekm, @wenzhixin, @djhvscf

I used the grouping example, and in a table with the information on the page it works as the template.

But in a dynamic way, I can not and always get the error: 

TypeError: Can not read property 'forEach' of undefined.

I've tried in several ways and I can not understand why the error. I have everything updated in the latest versions, so I do not think it's compatibility.

https://jsfiddle.net/SourceForge/za8bdfcp/1/

(I just was not able to link the actual data in jsfiddle, but I left a preview in the data.json)
